lib/client.ml

```diff
@@ -1,75 +1,148 @@
   in
   let msg_size =
     Encoding.read_size_header size_buffer + Encoding.size_header_length in
   let msg_buffer = Bytes.create msg_size in
   (* Now that we have read the header and the message size, we can read the message *)
-  let%lwt _, addr = recvfrom !client.socket msg_buffer 0 msg_size [] in
-  Lwt_mutex.unlock !client.recv_mutex;
-  Lwt.return (msg_buffer, Peer.from_socket_address addr)
+  let%lwt _ = recvfrom client_socket msg_buffer 0 msg_size [] in
+  let message = Encoding.unpack Message.bin_read_t msg_buffer in
+  Mutex.unlock !client.socket;
+  Lwt.return message
+
+let request client request recipient =
+  let%lwt message = create_request client recipient request in
+  let%lwt () = send_to client message in
+  let condition_var = Lwt_condition.create () in
+  Hashtbl.add !client.request_table message.id condition_var;
+  Lwt_condition.wait condition_var
+
+let broadcast_request client req recipients =
+  List.map (request client req) recipients
+
+let route client router msg =
+  let open Message in
+  let msg = router msg in
+  Inbox.push !client.inbox msg.category msg
 
-let serve client msg_handler =
+(* Signals a waiting request with its corresponding response
+   if it exists. Otherwise returns None. *)
+let handle_response request_table res =
+  let open Message in
+  let* res in
+  let* waiting_request = Hashtbl.find_opt request_table res.id in
+  Some (Lwt_condition.signal waiting_request res)
+
+(* Sever procedure:
+   1. Receive the next incoming message
+   2. Route the message
+   3. Grab the next response if it exists and send it to the request waiting for it
+   4. Grab the next request if it exists and send it to the message handler along with the
+     client's state
+   5. Send the encoded response from the message handler to the requester *)
+let serve client router msg_handler =
   let rec server () =
-    let%lwt request, peer = recv_next client in
-    let%lwt () = Lwt_mutex.lock !client.state_mutex in
-    let response = msg_handler !client.state peer request in
-    let%lwt () = send_to client response peer in
-    Lwt_mutex.unlock !client.state_mutex;
+    let%lwt message = recv_next client in
+    let%lwt () = route client router message in
+
+    let%lwt next_response = Inbox.next !client.inbox Message.Response in
+    let _ = handle_response !client.request_table next_response in
+
+    let%lwt request = Inbox.next !client.inbox Message.Request in
+    let%lwt () =
+      match request with
+      | Some request ->
+        let%lwt state = Mutex.lock !client.state in
+        let response =
+          request |> msg_handler state |> create_response client request in
+        let%lwt () = send_to client response in
+        Lwt.return (Mutex.unlock !client.state)
+      | None -> Lwt.return () in
+
     server () in
   Lwt.async server
 
-let init ~state ~msg_handler (address, port) =
+let init ~state ?(router = fun m -> m) ~msg_handler (address, port) =
```

Comment:
I am not sure about this `fun m -> m` does it mean 
> I need a router which is a function taking a `m` and returning the same `m`?

If yes, in Haskell it would have been `const m` is there the same in OCaml?
If no, would you mind explain me?

Reply:
It's the identity function yes, same as `const m`. We could use `Fun.id`but I'm not sure that's any more self explanatory. The purpose of `fun m -> m` here is that the router takes an incoming message as input, may change/decide its label, and may modify the payload before it's put into an inbox.